File: python/aurigraph_sdk/namespaces/wallet.py
"""Namespace for /api/v11/wallet endpoints -- wallet management and transfers."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any
from urllib.parse import quote

from aurigraph_sdk.models import Transaction, TransferReceipt, TransferRequest, WalletBalance

if TYPE_CHECKING:
    from aurigraph_sdk.client import AurigraphClient


class WalletApi:
    """Wallet management and transfer operations."""

    def __init__(self, client: AurigraphClient) -> None:
        self._client = client

    def get_balance(self, address: str) -> WalletBalance:
        """Get the balance for a wallet address.

        Args:
            address: The wallet address.

        Returns:
            Current wallet balance.
        """
        enc = quote(address, safe="")
        data = self._client._get(f"/wallet/{enc}/balance")
        return WalletBalance.model_validate(data)

    def transfer(self, request: TransferRequest) -> TransferReceipt:
        """Transfer tokens between wallets.

        Args:
            request: The transfer request details.

        Returns:
            Transfer receipt.
        """
        data = self._client._post("/wallet/transfer", request.model_dump())
        return TransferReceipt.model_validate(data)

    def get_history(self, address: str) -> list[Transaction]:
        """Get transaction history for a wallet address.

        Args:
            address: The wallet address.

        Returns:
            List of transactions for the address.
        """
        enc = quote(address, safe="")
        data = self._client._get(f"/wallet/{enc}/history")
        items: list[dict[str, Any]] = data if isinstance(data, list) else data.get("transactions", [])  # type: ignore[assignment]
        return [Transaction.model_validate(tx) for tx in items]

File: rust/src/namespaces/wallet.rs
//! Wallet management namespace for `/api/v11/wallet/*` endpoints.
//!
//! Provides balance queries, token transfers, and transaction history.
//!
//! # Example
//!
//! ```rust,no_run
//! # async fn example(client: &aurigraph_sdk::AurigraphClient) -> aurigraph_sdk::Result<()> {
//! let balance = client.wallet().get_balance("0xabc123").await?;
//! println!("Balance: {}", balance);
//!
//! let receipt = client.wallet().transfer("0xabc", "0xdef", 100.0).await?;
//! println!("Transfer: {}", receipt);
//! # Ok(())
//! # }
//! ```

use crate::client::AurigraphClient;
use crate::error::Result;

/// Wallet API namespace.
///
/// Obtained via [`AurigraphClient::wallet()`].
#[derive(Debug)]
pub struct WalletApi<'a> {
    client: &'a AurigraphClient,
}

impl<'a> WalletApi<'a> {
    pub(crate) fn new(client: &'a AurigraphClient) -> Self {
        Self { client }
    }

    /// Get wallet balance for an address.
    ///
    /// Calls `GET /api/v11/wallet/{address}/balance`.
    pub async fn get_balance(&self, address: &str) -> Result<serde_json::Value> {
        self.client
            .get(&format!("/wallet/{}/balance", address))
            .await
    }

    /// Transfer tokens between addresses.
    ///
    /// Calls `POST /api/v11/wallet/transfer`.
    pub async fn transfer(
        &self,
        from: &str,
        to: &str,
        amount: f64,
    ) -> Result<serde_json::Value> {
        let body = serde_json::json!({
            "fromAddress": from,
            "toAddress": to,
            "amount": amount,
        });
        self.client.post("/wallet/transfer", &body).await
    }

    /// Get transaction history for an address.
    ///
    /// Calls `GET /api/v11/wallet/{address}/history`.
    pub async fn get_history(&self, address: &str) -> Result<serde_json::Value> {
        self.client
            .get(&format!("/wallet/{}/history", address))
            .await
    }
}
